### PR TITLE
[WOR-86] Fix Azure workspace sharing

### PR DESCRIPTION
**Pull request by marctalbott**
Ticket: [WOR-86](https://broadworkbench.atlassian.net/browse/WOR-86)
* Route workspace acl updates for WSM workspaces to WSM
* Introduce WorkspaceAclManager to abstract out acl update logic specific to the workspace type
* Only do Requester Pays things if a workspace has a Google project. Requester Pays is unique to Google at the moment anyways, so Azure workspaces won't support Requester Pays anyways.

- [ ] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [ ] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [ ] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [ ] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [ ] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [ ] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [ ] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [ ] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 1]
```diff
@@ -1565,6 +1634,39 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
         ErrorReport(StatusCodes.BadRequest, "may not grant readers compute access")
       )
     }
+    if (
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Alternatively, we can not throw if someone tries to grant a reader or writer share/compute access on an McWorkspace and just silently not grant them that permission. I figured failing loudly was better than not granting the permissions that had been requested

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala [hunk 2]
```diff
@@ -217,4 +224,291 @@ class WorkspaceServiceUnitTests extends AnyFlatSpec with OptionValues with Mocki
...
+                  DateTime.now(),
+                  "creator@example.com",
+                  Map.empty
+        ).copy(workspaceType = workspaceType, googleProjectId = googleProjectId)
+      )
+    )
+    datasource
+  }
+
+  "getAcl" should "fetch policies from Sam for Rawls workspaces" in {
```
Comment (marctalbott) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala:319]: These tests mostly focus on checking that the right path is followed wrt Sam vs WSM. There are other unit tests in WorkspaceServiceSpec that focus more on the functionality of getAcl and updateAcl

File: core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala [hunk 3]
```diff
@@ -184,19 +184,22 @@ trait WorkspaceSupport {
...
 
   def getWorkspaceContext(workspaceName: WorkspaceName,
                           attributeSpecs: Option[WorkspaceAttributeSpecs] = None
-  ): Future[Workspace] =
-    dataSource.inTransaction { dataAccess =>
-      withWorkspaceContext(workspaceName, dataAccess, attributeSpecs) { workspaceContext =>
-        DBIO.successful(workspaceContext)
+  ): Future[Workspace] = {
+    for {
+      _ <- userEnabledCheck(ctx.userInfo)
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/util/WorkspaceSupport.scala:195]: Good catch 👍

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 4]
```diff
@@ -1448,58 +1512,73 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
...
+              case WorkspaceType.RawlsWorkspace =>
+                samDAO.addUserToPolicy(SamResourceTypeNames.workspace,
+                                       workspace.workspaceId,
+                                       policyName,
+                                       email,
+                                       ctx.userInfo
+                )
+              case WorkspaceType.McWorkspace =>
+                Future(
+                  Option(IamRole.fromValue(policyName.value.toUpperCase)).map { role =>
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: 🤔 This will mask the case of IamRole giving back null here won't it? We should throw an exception in that case, or at least log an error.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 5]
```diff
@@ -1448,58 +1512,73 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
...
+              case WorkspaceType.RawlsWorkspace =>
+                samDAO.removeUserFromPolicy(SamResourceTypeNames.workspace,
+                                            workspace.workspaceId,
+                                            policyName,
+                                            email,
+                                            ctx.userInfo
+                )
+              case WorkspaceType.McWorkspace =>
+                Future(
+                  Option(IamRole.fromValue(policyName.value.toUpperCase)).map { role =>
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Same question about the null case here.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 6]
```diff
@@ -1448,58 +1512,73 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
...
+        } yield {
+          val (invites, updates) = aclChanges.partition(acl => userToInvite.contains(acl.email))
+          sendACLUpdateNotifications(workspaceName,
+                                     updates
+          ) // we can blindly fire off this future because we don't care about the results and it happens async anyway
+          notificationDAO.fireAndForgetNotifications(inviteNotifications)
+          WorkspaceACLUpdateResponseList(updates,
+                                         invites,
+                                         Set.empty
+          ) // API_CHANGE: no longer return invitesUpdated because you technically can't do that anymore...
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Is this still a meaningful comment? I'm not grokking it's intent.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala (hunk 1 above)
Comment (aherbst-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Agreed, that is a programming error we should not mask.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 7]
```diff
@@ -1565,6 +1647,39 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
         ErrorReport(StatusCodes.BadRequest, "may not grant readers compute access")
       )
     }
+    if (
+      workspace.workspaceType.equals(WorkspaceType.McWorkspace) &&
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: It might be a bit cleaner to extract the common check for `WorkspaceType.McWorkspace` to a parent conditional rather than have it in all 3 conditionals here.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala [hunk 8]
```diff
@@ -47,29 +53,30 @@ class WorkspaceServiceUnitTests extends AnyFlatSpec with OptionValues with Mocki
     )
 
   def workspaceServiceConstructor(
-    datasource: SlickDataSource = mock[SlickDataSource],
-    methodRepoDAO: MethodRepoDAO = mock[MethodRepoDAO],
-    cromiamDAO: ExecutionServiceDAO = mock[ExecutionServiceDAO],
-    executionServiceCluster: ExecutionServiceCluster = mock[ExecutionServiceCluster],
+    datasource: SlickDataSource = mock[SlickDataSource](RETURNS_SMART_NULLS),
```
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala:56]: 👍

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 9]
```diff
@@ -1565,6 +1647,39 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
...
+      aclChanges.exists {
+        case WorkspaceACLUpdate(_, WorkspaceAccessLevels.Write, _, Some(true)) => true
+        case _                                                                 => false
+      }
+    ) {
+      throw new RawlsExceptionWithErrorReport(
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: Probably out of scope for this PR but food for thought: I've been trying to move toward throwing more targeted exception types rather than the catch-all RawlsExceptionWithErrorReport. It makes the code and associated tests more readable and I think it prevents us from missing edge cases in the tests (i.e., masking a real bug if a lower level piece of code throws `RawlsExceptionWithErrorReport`)

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala [hunk 10]
```diff
@@ -217,4 +224,298 @@ class WorkspaceServiceUnitTests extends AnyFlatSpec with OptionValues with Mocki
...
+    val datasource = mockDatasourceForAclTests(WorkspaceType.McWorkspace, workspaceId)
+    val samDAO = mockSamForAclTests()
+
+    val aclUpdates = Set(
+      WorkspaceACLUpdate(readerEmail, WorkspaceAccessLevels.Read, Option(true), Option(false))
+    )
+
+    val service =
+      workspaceServiceConstructor(datasource, samDAO = samDAO, workspaceManagerDAO = wsmDAO)(defaultRequestContext)
+    val exception = intercept[RawlsExceptionWithErrorReport] {
```
Comment (aherbst-broad) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala]: Same comment about `RawlsExceptionWithErrorReport` -- I'd like to be more certain we're not covering up an unrelated error condition here.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/HttpWorkspaceManagerDAO.scala [hunk 11]
```diff
@@ -201,4 +202,16 @@ class HttpWorkspaceManagerDAO(apiClientProvider: WorkspaceManagerApiClientProvid
         ),
       workspaceId
     )
+
+  override def getRoles(workspaceId: UUID, ctx: RawlsRequestContext) = getWorkspaceApi(ctx).getRoles(workspaceId)
+
+  override def grantRole(workspaceId: UUID, email: WorkbenchEmail, role: IamRole, ctx: RawlsRequestContext): Unit =
+    getWorkspaceApi(ctx).grantRole(
+      new GrantRoleRequestBody().memberEmail(email.value),
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/dataaccess/workspacemanager/HttpWorkspaceManagerDAO.scala:210]: Any way to get coverage on these in HttpWorkspaceManagerDaoSpec?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala [hunk 12]
```diff
@@ -1403,31 +1432,66 @@ class WorkspaceService(protected val ctx: RawlsRequestContext,
         )
       }
 
+    def getRawlsPolicies(workspaceId: UUID): Future[Set[(WorkbenchEmail, SamResourcePolicyName)]] =
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceService.scala]: `getRawlsPolicies` + `getWsmPolicies` both expose the same interface, there may be an abstraction to be had here -- `WorkspacePolicyManager` or some such thing, which would make this a bit more testable. I think this is probably fine for now though.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala (hunk 2 above)
Comment (aherbst-broad) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala:319]: There aren't any modifications to `WorkspaceServiceSpec` in this diff. Are you saying the existing coverage in that class is adequate given the additions to `WorkspaceServiceUnitTests` or is there more to be done?

**Review by aherbst-broad — APPROVED** [review]
Looking good. Had a few comments and some potential food for thought.

I pulled this down and ran against terra ui locally, it worked well.

**Review by ghost — APPROVED** [review]

**Comment by marctalbott** [comment]
Commits since last review start at [a6ef771](https://github.com/broadinstitute/rawls/pull/2038/commits/a6ef77130c0c73c7cabdb45fb2bb6734afa6255d)

File: core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala (hunk 2 above)
Comment (marctalbott) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceUnitTests.scala:319]: They seem pretty comprehensive to me (there's a bunch in `WorkspaceServiceSpec` and `WorkspaceApiServiceSpec`). They are more focused on validation and making sure `ACLUpdate`s get translated to the appropriate policy additions and removals. Some of that logic has been refactored in this PR, but the behavior should be the same.

Example of some of the tests that check all permutations of workspace ACL updates. These **are** using Sam, but they cover some of the more complicated parts of `updateAcl` that are shared between GCP and Azure workspaces like calculating required policy additions and removals - https://github.com/broadinstitute/rawls/blob/92443713d38a3b90500a94585f3f20e9bf9965b5/core/src/test/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceServiceSpec.scala#L1416

The new behavior that I'm most concerned with testing is the proper routing to Sam or WSM and translating WSM `RoleBindingList` to the equivalent Scala classes. I added a couple role bindings to the mocked WSM responses that should get ignored by Rawls (`DISCOVERER` and `APPLICATION`).

File: core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceAclManager.scala [hunk 13]
```diff
@@ -0,0 +1,38 @@
+package org.broadinstitute.dsde.rawls.workspace
+
+import org.broadinstitute.dsde.rawls.dataaccess.SamDAO
+import org.broadinstitute.dsde.rawls.model.{RawlsRequestContext, SamResourcePolicyName, Workspace, WorkspaceACL}
+import org.broadinstitute.dsde.workbench.model.WorkbenchEmail
+
+import java.util.UUID
+import scala.concurrent.{ExecutionContext, Future}
+
+trait WorkspaceAclManager {
```
Comment (aherbst-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/workspace/WorkspaceAclManager.scala:10]: 👍 I really like this refactoring

**Review by aherbst-broad — APPROVED** [review]
Looks good, thanks for taking on that refactoring.
